### PR TITLE
Add rsa pss rsae for tls12

### PR DESCRIPTION
**Pull request by yuhaoth**
When
    - GnuTLS/OpenSSL is configured as tls12 server with rsa key
    - `mebedTLS` is configured as hybrid mode.
    - The order is `rsa_pss_rsae_*`, `rsa_pkcs1_*`.
GnuTLS/OpenSSL will return `rsa_pss_rsae_*` which are not supported by TLS 1.2 in mbedTLS.

## Status
**READY**

### CODE REVIEW

**Comment by yuhaoth** [comment]
except armcc in OpenCI , others pass

File: library/ssl_tls12_client.c [hunk 1]
```diff
@@ -2402,14 +2342,31 @@ static int ssl_parse_server_key_exchange( mbedtls_ssl_context *ssl )
...
+#endif /* MBEDTLS_SSL_KEEP_PEER_CERTIFICATE */
+
         /*
          * Handle the digitally-signed structure
          */
-        if( ssl_parse_signature_algorithm( ssl, &p, end,
-                                           &md_alg, &pk_alg ) != 0 )
+        MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, 2 );
+        sig_alg = MBEDTLS_GET_UINT16_BE( p, 0 );
+        if( mbedtls_ssl_tls13_get_pk_type_and_md_alg_from_sig_alg(
```
Comment (paul-elliott-arm) [inline library/ssl_tls12_client.c:2366]: Is it ok to be using a TLS13 function from the TLS12 codebase? Should`  mbedtls_ssl_tls13_get_pk_type_and_md_alg_from_sig_alg()` be moved into ssl_client if it works for both 1.3 and 1.2?

File: tests/ssl-opt.sh [hunk 2]
```diff
@@ -12552,6 +12552,37 @@ run_test    "TLS 1.3: Check client no signature algorithm, m->m" \
             1 \
             -c "select_sig_alg_for_certificate_verify:no suitable signature algorithm found"
 
+requires_openssl_tls1_3
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
+requires_config_enabled MBEDTLS_DEBUG_C
+requires_config_enabled MBEDTLS_SSL_CLI_C
+run_test    "TLS 1.2: Check rsa_pss_rsae compitable issue, m->O" \
```
Comment (paul-elliott-arm) [inline tests/ssl-opt.sh:12559]:
```suggestion
run_test    "TLS 1.2: Check rsa_pss_rsae compatibility issue, m->O" \
```
Nit

File: tests/ssl-opt.sh [hunk 3]
```diff
@@ -12552,6 +12552,37 @@ run_test    "TLS 1.3: Check client no signature algorithm, m->m" \
...
+            0 \
+            -c "Protocol is TLSv1.2" \
+            -c "HTTP/1.0 200 [Oo][Kk]"
+
+
+requires_gnutls_tls1_3
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
+requires_config_enabled MBEDTLS_DEBUG_C
+requires_config_enabled MBEDTLS_SSL_CLI_C
+run_test    "TLS 1.2: Check rsa_pss_rsae compitable issue, m->G" \
```
Comment (paul-elliott-arm) [inline tests/ssl-opt.sh:12575]:
```suggestion
run_test    "TLS 1.2: Check rsa_pss_rsae compatibility issue, m->G" \
```
Nit

File: library/ssl_tls.c [hunk 4]
```diff
@@ -4171,6 +4159,18 @@ static uint16_t ssl_preset_default_sig_algs[] = {
     MBEDTLS_TLS1_3_SIG_RSA_PSS_RSAE_SHA256,
 #endif /* MBEDTLS_X509_RSASSA_PSS_SUPPORT && MBEDTLS_SHA256_C */
 
+#if defined(MBEDTLS_RSA_C) &&  defined(MBEDTLS_SHA512_C)
```
Comment (paul-elliott-arm) [inline library/ssl_tls.c:4162]: Doesn't this (and the whole PR) make the comment prior to this invalid?

File: ChangeLog.d/fix-rsa-pss-rsae-compitable-issue.txt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+Bugfix
```
Comment (paul-elliott-arm) [inline ChangeLog.d/fix-rsa-pss-rsae-compitable-issue.txt:1]: This doesn't describe to an end user what has been done here. For starters it is referring to internal code, and for seconds this describes the situation prior to this fix.
Would something more like "Add RSA PSA RSAE support for TLS1.2 client" be more descriptive?
Also, is this a bugfix or a new feature?

**Review by paul-elliott-arm — CHANGES_REQUESTED** [review]

File: library/ssl_tls12_client.c (hunk 1 above)
Comment (yuhaoth) [reply library/ssl_tls12_client.c:2366]: renamed

File: library/ssl_tls.c (hunk 4 above)
Comment (yuhaoth) [reply library/ssl_tls.c:4162]: comment is removed

**Comment by yuhaoth** [comment]
Rebased to resolve conflict.

**Review by paul-elliott-arm — DISMISSED** [review]
LGTM

File: library/ssl_tls.c [hunk 6]
```diff
@@ -4205,6 +4208,9 @@ static uint16_t ssl_tls12_preset_default_sig_algs[] = {
 #if defined(MBEDTLS_ECDSA_C)
     MBEDTLS_SSL_TLS12_SIG_AND_HASH_ALG( MBEDTLS_SSL_SIG_ECDSA, MBEDTLS_SSL_HASH_SHA384 ),
 #endif
+#if defined(MBEDTLS_X509_RSASSA_PSS_SUPPORT)
+    MBEDTLS_TLS1_3_SIG_RSA_PSS_RSAE_SHA384,
```
Comment (xkqian) [inline library/ssl_tls.c:4212]: Why we use the TLS13 macros in the guard of "#if defined(MBEDTLS_SSL_PROTO_TLS1_2)"?

File: library/ssl_tls.c [hunk 7]
```diff
@@ -4213,6 +4219,9 @@ static uint16_t ssl_tls12_preset_default_sig_algs[] = {
 #if defined(MBEDTLS_ECDSA_C)
     MBEDTLS_SSL_TLS12_SIG_AND_HASH_ALG( MBEDTLS_SSL_SIG_ECDSA, MBEDTLS_SSL_HASH_SHA256 ),
 #endif
+#if defined(MBEDTLS_X509_RSASSA_PSS_SUPPORT)
+    MBEDTLS_TLS1_3_SIG_RSA_PSS_RSAE_SHA256,
```
Comment (xkqian) [inline library/ssl_tls.c:4223]: ditto

File: library/ssl_tls.c [hunk 8]
```diff
@@ -4197,6 +4197,9 @@ static uint16_t ssl_tls12_preset_default_sig_algs[] = {
 #if defined(MBEDTLS_ECDSA_C)
     MBEDTLS_SSL_TLS12_SIG_AND_HASH_ALG( MBEDTLS_SSL_SIG_ECDSA, MBEDTLS_SSL_HASH_SHA512 ),
 #endif
+#if defined(MBEDTLS_X509_RSASSA_PSS_SUPPORT)
+    MBEDTLS_TLS1_3_SIG_RSA_PSS_RSAE_SHA512,
```
Comment (xkqian) [inline library/ssl_tls.c:4201]: ditto

File: library/ssl_tls.c (hunk 6 above)
Comment (yuhaoth) [reply library/ssl_tls.c:4212]: It is a backward compatible improvement. And the name is an IANA value come from RFC8446. 
The signature algorithm will be sent by OpenSSL server. I do not think we should redefine it.

File: library/ssl_tls.c (hunk 7 above)
Comment (yuhaoth) [reply library/ssl_tls.c:4223]: ditto

File: library/ssl_tls.c (hunk 8 above)
Comment (yuhaoth) [reply library/ssl_tls.c:4201]: ditto.

File: library/ssl_tls.c (hunk 6 above)
Comment (xkqian) [reply library/ssl_tls.c:4397]: It seems still wired, but I am ok with this workaround.

File: library/ssl_tls12_client.c [hunk 9]
```diff
@@ -2398,14 +2338,31 @@ static int ssl_parse_server_key_exchange( mbedtls_ssl_context *ssl )
...
+
         /*
          * Handle the digitally-signed structure
          */
-        if( ssl_parse_signature_algorithm( ssl, &p, end,
-                                           &md_alg, &pk_alg ) != 0 )
+        MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, 2 );
+        sig_alg = MBEDTLS_GET_UINT16_BE( p, 0 );
+        if( mbedtls_ssl_get_pk_type_and_md_alg_from_sig_alg(
+                                            sig_alg, &pk_alg, &md_alg ) != 0 &&
```
Comment (xkqian) [inline library/ssl_tls12_client.c:2363]:
```suggestion
                                sig_alg, &pk_alg, &md_alg ) != 0 &&
```
Better vertical allgined the `&&`

**Review by xkqian — CHANGES_REQUESTED** [review]

**Review by xkqian — APPROVED** [review]
LGTM

**Review by paul-elliott-arm — APPROVED** [review]
LGTM

**Pull request merged**